### PR TITLE
Keep reference to owner of data in Python Tensor and TensorList

### PR DESCRIPTION
**Pull request by klecki**
Signed-off-by: Krzysztof Lecki <klecki@nvidia.com>
#### Why we need this PR?
*Pick one, remove the rest*
- It fixes a bug with keeping alive the data wrapped by Tensor[List] in Python.

#### What happened in this PR?
*Fill relevant points, put NA otherwise. Replace anything inside []*
 - What solution was applied:
When creating the Tensor[List]Cpu/Gpu in Python from
Python-originated data, we only wrap the pointer to
allocation efectivelly creating a view. We need to indicate
that the original buffer needs to be kept alive, also when
the Tensor[List] is consumed by C++ backend and its Python
counterpart destroyed.

We store the copy of py::buffer/py::object in the custom deleter
allowing for it to be propagated with the data and
the ref count to be updated as needed.

The API was adjusted to set the shape if it's not empty.

 - Affected modules and functionalities:
     Tensor[List] Python wrappers, TensorList API
 - Key points relevant for the review:
     
 - Validation and testing:
     Test added
 - Documentation (including examples):
     Docs adjusted for TL.

**JIRA TASK**: *[Use DALI-XXXX or NA]*

### CODE REVIEW

**Comment by klecki** [comment]
!build

File: dali/pipeline/data/tensor_list.h [hunk 1]
```diff
@@ -251,8 +251,7 @@ class DLL_PUBLIC TensorList : public Buffer<Backend> {
    * if the size of the allocation is zero, the TensorList is reset to
    * a default state and is NOT marked as sharing data.
    *
-   * After wrapping the allocation, the TensorLists size is set to 0,
-   * and its type is reset to NoType.
+   * After wrapping the allocation, the shape and type is set if were provided non-empty.
```
Comment (mzient) [inline dali/pipeline/data/tensor_list.h:254]: Something's wrong with this sentence.
```suggestion
   * The size of the tensor list is calculated based on shape and type or reset to 0
   * if the shape is empty or the type is DALI_NO_TYPE.
```

File: dali/python/backend_impl.cc [hunk 2]
```diff
@@ -343,9 +344,10 @@ void ExposeTensor(py::module &m) {
           auto t = std::make_unique<Tensor<CPUBackend>>();
           t->set_pinned(is_pinned);
           TypeInfo type = TypeFromFormatStr(info.format);
-          t->ShareData(info.ptr, bytes, type);
+          // Keep the copy of input buffer ref in the deleter, so it's refcount is increased
```
Comment (mzient) [inline dali/python/backend_impl.cc:347]:
```suggestion
          // Keep a copy of the input buffer ref in the deleter, so its refcount is increased
```

File: dali/python/backend_impl.cc [hunk 3]
```diff
@@ -614,9 +616,10 @@ void ExposeTensorList(py::module &m) {
         auto t = std::make_shared<TensorList<CPUBackend>>();
         t->set_pinned(false);
         TypeInfo type = TypeFromFormatStr(info.format);
-        t->ShareData(info.ptr, bytes, type);
+        // Keep the copy of input buffer ref in the deleter, so it's refcount is increased
```
Comment (mzient) [inline dali/python/backend_impl.cc:619]:
```suggestion
        // Keep a copy of the input buffer ref in the deleter, so its refcount is increased
```

File: dali/test/python/test_tensor_cupy.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        np.testing.assert_array_equal(np.array(d), np.zeros((4, 4)) + i)
+
+def test_tensor_list_cpu_from_numpy():
+    def create_tmp(idx):
+        a = np.zeros((4, 4)) + idx
+        return dali.tensors.TensorListCPU(a, "")
+    out = [create_tmp(i) for i in range(4)]
+    for i, d in enumerate(out):
+        np.testing.assert_array_equal(d.as_array(), np.zeros((4, 4)) + i)
+
```
Comment (mzient) [inline dali/test/python/test_tensor_cupy.py:42]: I'd like to see a test that creates a TensorListCPU as a temporary object and checks the contents of the tensors taken from it with []:
```Python
tensor = [t for t in make_tensor_list()]
for i, t in enumerate(tensors):
  np.testing.assert_array_equal(np.array(t), np.full(t.shape(), i))
```

File: dali/test/python/test_tensor_cupy.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+# Without this behaviour there was observable bug with creating several temporary
+# buffers in the loop and DALI not tracking references to them
+
+def test_tensor_cpu_from_numpy():
+    def create_tmp(idx):
+        a = np.zeros((4, 4)) + idx
+        return dali.tensors.TensorCPU(a, "")
+    out = [create_tmp(i) for i in range(4)]
+    for i, d in enumerate(out):
+        np.testing.assert_array_equal(np.array(d), np.zeros((4, 4)) + i)
```
Comment (mzient) [inline dali/test/python/test_tensor_cupy.py:33]:
```suggestion
        np.testing.assert_array_equal(np.array(d), np.full((4, 4), i))
```

File: dali/test/python/test_tensor_cupy.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+def test_tensor_list_cpu_from_numpy():
+    def create_tmp(idx):
+        a = np.zeros((4, 4)) + idx
+        return dali.tensors.TensorListCPU(a, "")
+    out = [create_tmp(i) for i in range(4)]
+    for i, d in enumerate(out):
+        np.testing.assert_array_equal(d.as_array(), np.zeros((4, 4)) + i)
```
Comment (mzient) [inline dali/test/python/test_tensor_cupy.py:41]:
```suggestion
        np.testing.assert_array_equal(d.as_array(), np.full((4, 4), i))
```

File: dali/test/python/test_tensor_cupy.py [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+def test_tensor_gpu_from_cupy():
+    def create_tmp(idx):
+        a = np.zeros((4, 4)) + idx
+        a_gpu = cp.array(a, dtype=a.dtype)
+        return dali.tensors.TensorGPU(a_gpu, "")
+    out = [create_tmp(i) for i in range(4)]
+    for i, d in enumerate(out):
+        np.testing.assert_array_equal(np.array(d.as_cpu()), np.zeros((4, 4)) + i)
```
Comment (mzient) [inline dali/test/python/test_tensor_cupy.py:50]:
```suggestion
        np.testing.assert_array_equal(np.array(d.as_cpu()), np.full((4, 4), i))
```

File: dali/test/python/test_tensor_cupy.py [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+
+def test_tensor_list_gpu_from_cupy():
+    def create_tmp(idx):
+        a = np.zeros((4, 4)) + idx
+        a_gpu = cp.array(a, dtype=a.dtype)
+        return dali.tensors.TensorListGPU(a_gpu, "")
+    out = [create_tmp(i) for i in range(4)]
+    for i, d in enumerate(out):
+        np.testing.assert_array_equal(d.as_cpu().as_array(), np.zeros((4, 4)) + i)
```
Comment (mzient) [inline dali/test/python/test_tensor_cupy.py:59]:
```suggestion
        np.testing.assert_array_equal(d.as_cpu().as_array(), np.full((4, 4), i))
```

File: dali/test/python/test_tensor_cupy.py [hunk 9]
```diff
@@ -0,0 +1,59 @@
+# Copyright (c) 2021, NVIDIA CORPORATION. All rights reserved.
```
Comment (JanuszL) [inline dali/test/python/test_tensor_cupy.py:1]: Why not move this tests to test_backend_impl_gpu.py ?

File: dali/pipeline/data/tensor_list.h (hunk 1 above)
Comment (klecki) [reply dali/pipeline/data/tensor_list.h:254]: done

File: dali/python/backend_impl.cc (hunk 2 above)
Comment (klecki) [reply dali/python/backend_impl.cc:347]: done

File: dali/python/backend_impl.cc (hunk 3 above)
Comment (klecki) [reply dali/python/backend_impl.cc:619]: done

File: dali/test/python/test_tensor_cupy.py (hunk 9 above)
Comment (klecki) [reply dali/test/python/test_tensor_cupy.py:1]: Done

File: dali/test/python/test_tensor_cupy.py (hunk 5 above)
Comment (klecki) [reply dali/test/python/test_tensor_cupy.py:33]: Done

File: dali/test/python/test_backend_impl_gpu.py [hunk 10]
```diff
@@ -216,3 +217,51 @@ def check_squeeze(shape, dim, in_layout, expected_out_layout):
...
+
+def test_tensor_list_cpu_from_numpy():
+    def create_tmp(idx):
+        a = np.full((4, 4), idx)
+        return tensors.TensorListCPU(a, "")
+    out = [create_tmp(i) for i in range(4)]
+    for i, tl in enumerate(out):
+        for j in range(4):
+            np.testing.assert_array_equal(np.array(tl[j]), np.full(tl[j].shape(), i))
+        np.testing.assert_array_equal(tl.as_array(), np.full((4, 4), i))
```
Comment (JanuszL) [inline dali/test/python/test_backend_impl_gpu.py:246]: Numpy one can go just to test_backend_impl.py.
I wonder if it is worth to create a similar test for DLPack as well (we have tests for DLPack in test_backend_impl_torch_dlpack.py) but not for this scenario.

**Review by JanuszL — APPROVED** [review]

File: dali/test/python/test_backend_impl_gpu.py (hunk 10 above)
Comment (klecki) [reply dali/test/python/test_backend_impl_gpu.py]: Added tests for dlpack. I would prefer to keep them grouped and not spread in several files.

**Comment by klecki** [comment]
!build

File: dali/test/python/test_backend_impl_gpu.py (hunk 10 above)
Comment (JanuszL) [reply dali/test/python/test_backend_impl_gpu.py]: Fair point

**Review by mzient — APPROVED** [review]

**Pull request merged**